packages/cli/src/evaluation/test-runner/test-runner.service.ee.ts

```diff
@@ -0,0 +1,146 @@
+import { parse } from 'flatted';
+import type { IPinData, IWorkflowExecutionDataProcess } from 'n8n-workflow';
+import assert from 'node:assert';
+import { Container, Service } from 'typedi';
+
+import { ActiveExecutions } from '@/active-executions';
+import type { ExecutionEntity } from '@/databases/entities/execution-entity';
+import type { User } from '@/databases/entities/user';
+import type { WorkflowEntity } from '@/databases/entities/workflow-entity';
+import { ExecutionRepository } from '@/databases/repositories/execution.repository';
+import { WorkflowRepository } from '@/databases/repositories/workflow.repository';
+import { NotFoundError } from '@/errors/response-errors/not-found.error';
+import { TestDefinitionService } from '@/evaluation/test-definition.service.ee';
+import type { IExecutionDb, IExecutionResponse } from '@/interfaces';
+import { WorkflowRunner } from '@/workflow-runner';
+
+/**
+ * This service orchestrates the running of test cases.
+ * It uses the test definitions to find
+ * past executions, creates pin data from them,
+ * and runs the workflow-under-test with the pin data.
+ * TODO: Evaluation workflows
+ * TODO: Node pinning
+ * TODO: Collect metrics
+ */
+@Service()
+export class TestRunnerService {
+	constructor(
+		private readonly testDefinitionsService: TestDefinitionService,
+		private readonly workflowRepository: WorkflowRepository,
+		private readonly workflowRunner: WorkflowRunner,
+		private readonly executionRepository: ExecutionRepository,
+	) {}
+
+	/**
+	 * Creates a pin data object from the past execution data
+	 * for the given workflow.
+	 * For now, it only pins trigger nodes.
+	 * @param workflow
+	 * @param execution
+	 * @private
+	 */
+	private createPinDataFromExecution(
+		workflow: WorkflowEntity,
+		execution: ExecutionEntity,
+	): IPinData {
+		const executionData = parse(execution.executionData.data) as IExecutionResponse['data'];
+
+		const triggerNodes = workflow.nodes.filter((node) => /trigger$/i.test(node.type));
+
+		const pinData = {} as IPinData;
+
+		for (const triggerNode of triggerNodes) {
+			const triggerData = executionData.resultData.runData[triggerNode.name];
+			if (triggerData?.[0]?.data?.main?.[0]) {
+				pinData[triggerNode.name] = triggerData[0]?.data?.main?.[0];
+			}
+		}
+
+		return pinData;
+	}
+
+	/**
+	 * Runs a test case with the given pin data.
+	 * Waits for the workflow under test to finish execution.
+	 * @param workflow
+	 * @param testCase
+	 * @param userId
+	 * @private
+	 */
+	private async runTestCase(
+		workflow: WorkflowEntity,
+		testCase: IPinData,
```

Comment:
The parameter name and type have bit of a mismatch. Would `testData` be accurate? Or `testInputData`? Or something like that